opba-banking-protocol-facade/src/main/java/de/adorsys/opba/protocol/facade/services/ServiceContextProvider.java

```diff
@@ -58,38 +79,79 @@
+                                               EncryptionService encryptionService,
+                                               KeyDTO keyDTO) {
+        String encryptedContext = new String(encryptionService.encrypt(MAPPER.writeValueAsBytes(facadeServiceable)));
+        return new ContextWithKey(encryptedContext, keyDTO);
+    }
+
     private <T extends FacadeServiceableGetter> ServiceSession extractOrCreateServiceSession(
             T request,
-            AuthSession authSession
+            AuthSession authSession,
+            Supplier<ContextWithKey> encryptedContext
     ) {
         if (null != authSession) {
             return authSession.getParent();
         } else {
-            return findServiceSessionByIdOrCreate(request);
+            return findServiceSessionByIdOrCreate(request, encryptedContext);
         }
     }
 
     @NotNull
     @SneakyThrows
-    private <T extends FacadeServiceableGetter> ServiceSession findServiceSessionByIdOrCreate(T request) {
-        if (null != request.getFacadeServiceable().getServiceSessionId()) {
-            Optional<ServiceSession> existingSession = serviceSessions
-                    .findById(request.getFacadeServiceable().getServiceSessionId());
-            if (existingSession.isPresent()) {
-                return existingSession.get();
+    private <T extends FacadeServiceableGetter> ServiceSession findServiceSessionByIdOrCreate(
+            T request,
+            Supplier<ContextWithKey> contextWithKey
+    ) {
+        FacadeServiceableRequest facadeServiceable = request.getFacadeServiceable();
+        UUID serviceSessionId = facadeServiceable.getServiceSessionId();
+
+        ServiceSession session = new ServiceSession();
+        if (null != serviceSessionId) {
+            session.setId(serviceSessionId);
+        }
+
+        session.setContext(contextWithKey.get().getEncryptedContext());
+        session.setFintechOkUri(facadeServiceable.getFintechRedirectUrlOk());
+        session.setFintechNokUri(facadeServiceable.getFintechRedirectUrlNok());
+        session.setSecretKey(secretKeyService.encrypt(contextWithKey.get().getKeyDTO().getKey()));
+        session.setAlgo(properties.getAlgorithm());
+        session.setSalt(contextWithKey.get().getKeyDTO().getSalt());
+        session.setIterCount(properties.getIterationCount());
+        return serviceSessions.save(session);
+    }
+
+    private KeyDTO getSessionSecretKey(UUID serviceSessionId, String sessionPassword) {
+
+        // new secret key
+        if (null == serviceSessionId) {
+            if (Strings.isNullOrEmpty(sessionPassword)) {
+                throw new RuntimeException("No password. Can't generate secret key");
             }
+            byte[] salt = getNewSalt(properties.getSaltLength());
+            byte[] key = secretKeyService.generateKey(sessionPassword, salt);
+            return new KeyDTO(key, salt);
```

Comment:
Given the importance of encryption, I suggest creating:
KeyDto that simply stores key bytes[]
that extends KeyDto
KeyAndSaltDto that store both key and salt bytes

Reply:
done